Improve support for user-supplied packer to FreeTypeFontGenerator

RobRendell opened the pull request:
This is a fresh pull request from a fresh branch with the changes from #2993, because my merging and rebasing got messed up.

NathanSweet commented on extensions/gdx-freetype/src/com/badlogic/gdx/graphics/g2d/freetype/FreeTypeFontGenerator.java:
```diff
@@ -549,29 +531,17 @@ Glyph createGlyph (char c, FreeTypeBitmapFontData data, FreeTypeFontParameter pa
-			throw new IllegalStateException("Packer was not able to insert glyph into a page: " + name);
-
-		glyph.page = pageIndex;
+		glyph.page = packer.getPages().size - 1;
```
The packed pixmap could have been put on any page. The page index needs to be gotten from the packer somehow.

RobRendell replied:
I'm not sure if the packing algorithm has changed (I didn't change it), but by my reading in the current version it will always be the last page.  A new Pixmap passed to `pack` will only check `currPage` for space, and if it doesn't fit a new page is created, `currPage` is updated to point to the new page, and it tries to pack again.  I agree that it would result in a more efficient pack if `pack` scanned all pages for space for its Pixmap, but I don't believe that's how it currently works.

NathanSweet replied:
Ah, you're right (the packer isn't my code). We can make the assumption for now. Later we can do it another way if needed (maybe the packer returns something that has a rectangle and metadata).

NathanSweet commented on gdx/src/com/badlogic/gdx/graphics/g2d/PixmapPacker.java:
```diff
@@ -155,39 +152,109 @@ public PixmapPacker (int width, int height, Format format, int padding, boolean
-	 * Inserts the given {@link Pixmap}. You can later on retrieve the images position in the output image via the supplied name
+	 * Inserts the given {@link Pixmap} with no name.  You must use the returned Rectangle to later access the
+	 * Pixmap's location in the appropriate page (which will be the last page in the PixmapPacker at the time this
+	 * method returns).
```
Need to fix up javadocs, may not be last page.

NathanSweet commented on gdx/src/com/badlogic/gdx/graphics/g2d/PixmapPacker.java:
```diff
@@ -155,39 +152,109 @@ public PixmapPacker (int width, int height, Format format, int padding, boolean
+	 * @param image the image
+	 * @return Rectangle describing the area the pixmap was rendered to.
+	 * @throws RuntimeException in case the image did not fit due to the page size being too small or providing a duplicate name */
+	public synchronized Rectangle packDirectToTexture(String name, Pixmap image) {
```
If name can be null, the parameter javadoc should let this be known.

NathanSweet commented:
There are some good improvements in here. I wish usage was simpler. Mixing updating textures with the pixmap packing is something that feels like could be decoupled. A method or two for convenience is one thing, but it's started to permeate all of the packer API. I am thinking about merging this and then do some refactoring.

RobRendell commented:
The other option would be to have a `directToTexture` flag on the PixmapPacker that could be turned on and off, instead of the new `packDirectToTexture` methods.

Re: refactoring: please, feel free to hack it around as much as you like :)  I just wanted the ability to call `FreeTypeFontGenerator.generateFont()` with a user-supplied PixmapPacker, as raised in issue #2980.  I'm enjoying contributing to the project, but you're the boss :)

RobRendell commented:
Updated to HEAD, should be auto-mergeable again.

I've updated the javadoc for pack(String, Pixmap) to explain how it works if `name` is null.  Given that PixmapPacker does always pack to the last page, I think that's all the comments addressed.

RobRendell commented:
I just merged this branch with HEAD again.

NathanSweet commented:
I merged your PR and refactored things a bit. Can you take a look? I'm not super happy with it all, but it's better than before we started hacking on it.

RobRendell commented:
It looks good to me.  One thing I noticed: you removed the "dirty" test in PixmapPacker line 161:

Was
```
		if (directToTexture && currPage.texture != null && !currPage.textureNeedsRefresh && !duplicateBorder) {
```
now
```
		if (packToTexture && current.texture != null && !duplicateBorder) {
```

My reasoning behind testing `curPage.textureNeedsRefresh` (now `currPage.dirty`) is that if the page is marked as dirty, it's going to have to be re-loaded before it's used anyway.  The `page.image` pixmap is going to be updated with the new pixmap anyway - there's no point in incurring the additional cost of updating the texture directly as well, given that the whole texture is going to be re-loaded from `page.image` later anyway.

NathanSweet commented:
Good catch! Fixing.

loogie commented:
Hello!

 

You have a new message, please read <http://ks.mpt.ru/begin.php?n1n>

 

loogie@basicelementdesign.com
